tests/upgrades/test_remoteexecution.py

```diff
@@ -0,0 +1,165 @@
+    """The test class contains pre-upgrade and post-upgrade scenarios to test if
+        remote execution working with external capsule.
+
+        Test Steps:
+
+        1. Before Satellite upgrade, Create VM,Subnet,install katello-ca,
+           register, add_ssh_key of external capsule,run a rex job on client vm
+        2. Upgrade satellite/capsule.
+        3. Run a rex Job again with same client vm.
+        4. Check if rex job still getting success.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        cls.libvirt_vm = os.environ.get('LIBVIRT_HOSTNAME')
+        cls.org = entities.Organization(id=1).read()
+        cls.bridge = os.environ.get('BRIDGE')
+        cls.subnet = os.environ.get('SUBNET')
+        cls.gateway = os.environ.get('GATEWAY')
+        cls.netmask = os.environ.get('NETMASK')
+        cls.vm_domain_name = os.environ.get('VM_DOMAIN')
+        cls.vm_domain = entities.Domain().search(query={'search': 'name="{}"'
+                                                 .format(cls.vm_domain_name)})
+        cls.proxy_name = os.environ.get('RHEV_CAP_HOST')
+
+    def _get_succeeded_count(self, id=None):
+        if id:
+            job_status = entities.JobInvocation(id=id).read()
+            return job_status
+
+    def _validated_succeeded_count(self, id=None):
+        """ Validate the rex job success count """
+        if id:
+            wait_for(
+                lambda: isinstance(self._get_succeeded_count(id=id).succeeded,
+                                   int) and self._get_succeeded_count(id=id).succeeded > 0,
+                timeout=30,
+                delay=2,
+                logger=self.logger
+            )
+        return self._get_succeeded_count(id=id)
+
+    def _vm_cleanup(self, hostname=None):
+        """ Cleanup the VM from provisioning server"""
+        if hostname:
+            vm = VirtualMachine(
+                hostname=hostname,
+                target_image=hostname,
+                provisioning_server=self.libvirt_vm,
+                distro=DISTRO_RHEL7,
+                )
+            vm._created = True
+            vm.destroy()
+
+    @pre_upgrade
+    def test_pre_scenario_remoteexecution_external_capsule(self):
+        """Pre-upgrade scenario that creates the client VM with
+        Subnet,install katello-ca pakcage, register, add_ssh_key of
+        external capsule and run a rex job on it.
+
+        :id: preupgrade-261dd2aa-be01-4c34-b877-54b8ee346561
+
+        :steps:
+            1. Create Subnet
+            2. Create client vm
+            3. Install katello-ca package and register to Satellite host
+            4. add rex ssh_key of external capsule on client vm
+            5. run the rex job on client vm
+
+        :expectedresults:
+            1. client vm should create with pre-required details and
+            2. rex job should run on client vm.
+        """
+        try:
+            sn = entities.Subnet(
```

Comment:
What if this subnet is already created ? first check if that vm is already created.